### ISSUE TITLE
`drv$list_shared_items()` returns list of `ms_object` and not list of `ms_drive_items`

### ISSUE THREAD

**Issue by mstackhouse** [Azure/Microsoft365R#206]
Currently I'm trying to access shared a shared folder, so in order to find the shared folder item I'm using the method `drv$list_shared_items()`. The list that's returned is a list of `ms_object` objects.

Prior to version, I could use that item directly to access files from within the shared folder using `folder_item$get_item("path")`, but now that the items are classed as `ms_object` the `get_item()` method is not available.

In the documentation, it states:

> Because of how the Graph API handles access to shared items linked in the root, you cannot directly access subitems of shared folders via the drive get_item method, like this: `drv$get_item("shared_folder/path/to/file")`. Instead, get the item into its own object, and use its get_item method: `drv$get_item("shared_folder")$get_item("path/to/file")`

Unfortunately I can't get this to work, because:
- I can't find an applicable way to enter the "shared_folder" path, because any path that I provide is pointing into my own OneDrive folder as the root path, and
- I can't find any ItemID within the ms_object properties that is accessible using `drv$get_item()`. Any ID that I try returns a 404.

**Comment by dcomputing** [comment]
The public version 2.4.0 does not return a list of ms_drive_items as per NEWS: The list_shared_items()/list_shared_files() method for drives now always returns a list of drive item objects [https://cran.r-project.org/web/packages/Microsoft365R/news/news.html](url)

As seen in the 2.4.0 version of the function below, it returns a vector based on the drive item objects and this causes all code depending on a list of drive item objects from list_shared_items() to fail since drive item object functions are no longer available. Whatever happened to issue #189 which is supposed to fix this problem?

list_shared_items=function(allow_external=TRUE, filter=NULL, n=Inf, pagesize=1000, info=NULL)
    {
        if(!is.null(info) && info != "items")
            warning("Ignoring 'info' argument, returning a list of drive items")

        opts <- list(`$top`=pagesize)
        if(allow_external)
            opts$allowExternal <- "true"
        if(!is.null(filter))
            opts$`filter` <- filter
        children <- self$do_operation("sharedWithMe", options=opts, simplify=FALSE)

        # get file list as a data frame, or return the iterator immediately if n is NULL
        out <- extract_list_values(self$get_list_pager(children), n)
        names(out) <- sapply(out, function(obj) obj$properties$name)
        out
    }

**Comment by mstackhouse** [comment]
I think you just sent this via a reply to the email, but just an FYI this
didn't update as a comment on the GitHub issue itself:
https://github.com/Azure/Microsoft365R/issues/206

**Comment by awosoga** [comment]
Experiencing this issue as well, has a workaround been proposed yet?

**Comment by msberends** [comment]
Sure, I’d like to help - but if @hongooi73 is not available to merge PR’s, how will this package progress? Including releases on CRAN?

**Comment by stephenashton-dhsc** [comment]
I probably won't be able to provide too much support - whilst I am reasonably okay with package development, I'm not skilled enough with the underlying API interface.

For package progress, I'm assuming that @hongooi73 or someone within the Azure organisation will have to change permissions to allow for a wider number of individuals to merge PRs. For releases on CRAN, I believe the standard process is for the current author to email CRAN formally passing authorship to another named individual.

**Comment by 3styleJam** [comment]
Thank you @msberends and @stephenashton-dhsc for your quick responses. I am not sure who the appropriate point of contact would be in the Azure organisation if @hongooi73 is unavailable, may require a general enquiry. What happens if an author of a package is unavailable and therefore unable to formally pass authorship? Is there a process where ownership can be transferred, or failing that grant merge (or other) permissions to other trusted contributors?

**Comment by stephenashton-dhsc** [comment]
For CRAN, their requirement is to "explain any change in the maintainer’s email address and if possible send confirmation from the previous address (by a separate email to CRAN-submissions@R-project.org) or explain why it is not possible." 

I haven't gone through this process myself, but some of it may depend on why the transfer is occurring (e.g. if @hongooi73 is no longer part of Microsoft).

Regarding reaching out to Azure, it appears as though they list the contact address for the organisation as [opensource@microsoft.com](mailto:opensource@microsoft.com).

**Comment by hongooi73** [comment]
Hello all, apologies for the non-response, I've been dealing with some personal issues. I'll be looking at the accumulated issues over the next few days and merging/fixing as required. Sorry again for not being able to get this done before.

**Comment by hongooi73** [comment]
Fixing shared files is difficult for me, as I don't have access to a business MS365 account (as far as I know, everything works with a personal account). Can someone who has an account share a file for testing purposes?

**Comment by mstackhouse** [comment]
@hongooi73 if you give me an account name I'm happy to share a file with you. You can send to michaelstackhouse0614@gmail.com and I'll reach out from my business account.

**Comment by hongooi73** [comment]
@mstackhouse I've just sent you an email. Thanks!

**Comment by hongooi73** [comment]
Just to clarify, is this problem occurring with the CRAN version of the package, or the dev version on Github?

**Comment by 3styleJam** [comment]
> Just to clarify, is this problem occurring with the CRAN version of the package, or the dev version on Github?

I am experiencing this problem on the CRAN version, can't speak for the dev version as I haven't tried it.

**Comment by hongooi73** [comment]
Please try the Github version, as it makes changes to the shared object code.

**Comment by hongooi73** [comment]
Hmm.. it looks like Business OD/SP files shared with a personal account don't show up in the listing 🤔

**Comment by 3styleJam** [comment]
> Hmm.. it looks like Business OD/SP files shared with a personal account don't show up in the listing 🤔

I have tried the GitHub version, connected to a Business OneDrive and ran `drv$list_shared_items()`. I have then used `drv$list_shared_items()[[i]]` to grab the ith item, and `drv$list_shared_items()[[i]]$download()` successfully downloads the file. This is something I could not do on the CRAN version because there was no download method.. Also note that in my attempt I chose a file that was shared with me from another Business OneDrive, so I am not sure why it wouldn't work with a file shared from a Personal OneDrive.

**Comment by hongooi73** [comment]
What I mean is that on my personal MS365 account, the files that people have shared with me from a business account don't show up. But it sounds like the Github version fixes the issue of `list_shared_items` returning incorrect object types, is this correct?

**Comment by 3styleJam** [comment]
> What I mean is that on my personal MS365 account, the files that people have shared with me from a business account don't show up. But it sounds like the Github version fixes the issue of `list_shared_items` returning incorrect object types, is this correct?

Correct. The CRAN version returns R6 objects of class ms_object only (which does not have the download or get_item methods), while the GitHub version returns R6 objects of ms_drive_item and ms_object and those do have download, get_item and presumably the other methods expected for the ms_drive_item class. @mstackhouse please can you test your original code again with the GitHub version of Microsoft365R and let us know if it works or not?

**Comment by hongooi73** [comment]
I'm going to mark this as done. If people are still experiencing problems, feel free to reply to this thread.
